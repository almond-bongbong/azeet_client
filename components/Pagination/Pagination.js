import React, { useMemo } from 'react';
import range from 'lodash/range';
import takeWhile from 'lodash/takeWhile';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { blue } from 'style/mixin';
import { Icon } from 'antd';

const PaginationStyle = styled.div`
  display: block;
  color: #888;
  font-size: 0;
  text-align: center;
  .arrow {
    padding: 6px 0 4px;
    color: #999;
    vertical-align: middle;
    i {
      display: block;
      font-size: 13px;
    }
  }
  .numbers {
    display: inline-block;
    margin: 0 10px;
    font-size: 14px;
    vertical-align: middle;
  }
  button {
    display: inline-block;
    width: 30px;
    height: 30px;
    padding: 5px 0;
    border: 0;
    outline: 0;
    background: #fff;
    cursor: pointer;
    text-align: center;
    &:hover {
      background-color: #f2f2f2;
    }
  }
`;

const NumberStyle = styled.button`
  span {
    ${props => props.active && css`
      color: ${blue};
      text-decoration: underline;
    `}
  }
`;

const Pagination = ({
  current, displayItem, displayPage, total, onChange,
}) => {
  const pageCount = Math.ceil(total / displayItem) || 1;
  const displayStartPage = Math.floor((current - 1) / displayPage) * displayPage + 1;
  const displayEndPage = displayStartPage + displayPage - 1;
  const prevPage = displayStartPage > 1 ? displayStartPage - 1 : 1;
  const nextPage = displayEndPage < pageCount ? displayEndPage + 1 : pageCount;
  const numbers = useMemo(() => (
    takeWhile(range(displayStartPage, displayEndPage + 1), n => n <= pageCount)
      .map(i => (
        <NumberStyle
          key={i}
          type="button"
          className="num"
          active={i === current}
          onClick={() => onChange(i)}
        >
          <span>{i}</span>
        </NumberStyle>
      ))), [displayStartPage, displayEndPage, pageCount, current, onChange]);

  return (
    <PaginationStyle>
      <button title="첫 페이지로 이동" type="button" className="arrow" onClick={() => onChange(1)}>
        <Icon type="vertical-right" />
      </button>
      <button title={`${prevPage} 페이지로 이동`} type="button" className="arrow" onClick={() => onChange(prevPage)}>
        <Icon type="left" />
      </button>
      <div className="numbers">{numbers}</div>
      <button title={`${nextPage} 페이지로 이동`} type="button" className="arrow" onClick={() => onChange(nextPage)}>
        <Icon type="right" />
      </button>
      <button title="마지막 페이지로 이동" type="button" className="arrow" onClick={() => onChange(pageCount)}>
        <Icon type="vertical-left" />
      </button>
    </PaginationStyle>
  );
};

Pagination.propTypes = {
  current: PropTypes.number,
  displayItem: PropTypes.number,
  displayPage: PropTypes.number,
  total: PropTypes.number.isRequired,
  onChange: PropTypes.func,
};

Pagination.defaultProps = {
  current: 1,
  displayItem: 10,
  displayPage: 5,
  onChange: null,
};

export default Pagination;
