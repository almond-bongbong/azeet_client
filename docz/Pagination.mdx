---
name: Pagination
---

import { Playground, Props } from 'docz'
import Pagination from 'components/Pagination/Pagination'

# Pagination

<Props of={Pagination} />

## Basic usage

<Playground>
  <Pagination type="text" current={4} total={115} onChange={page => console.log(page)} />
</Playground>
