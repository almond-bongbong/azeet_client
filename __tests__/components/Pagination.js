import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Pagination from 'components/Pagination/Pagination';

describe('Pagination', () => {
  const activatedPageStyle = 'text-decoration: underline';

  it('matches snapshot', () => {
    const utils = render(<Pagination total={23} />);

    expect(utils.container).toMatchSnapshot();
  });

  it('current 2 page', () => {
    const { getByText, queryByText } = render(<Pagination current={2} total={24} />);

    getByText('1');
    expect(getByText('2')).toHaveStyle(activatedPageStyle);
    getByText('3');
    expect(queryByText('4')).toBeNull();
  });

  it('click to first page', () => {
    const onChange = jest.fn();
    const { getByTitle } = render(<Pagination current={3} total={51} onChange={onChange} />);
    const btnFirstPage = getByTitle('첫 페이지로 이동');

    fireEvent.click(btnFirstPage);
    expect(onChange).toBeCalledWith(1);
  });

  it('click to last page', () => {
    const onChange = jest.fn();
    const { getByTitle } = render(<Pagination current={3} total={51} onChange={onChange} />);
    const btnLastPage = getByTitle('마지막 페이지로 이동');

    fireEvent.click(btnLastPage);
    expect(onChange).toBeCalledWith(6);
  });

  it('click to start page', () => {
    const onChange = jest.fn();
    const { getByTitle } = render(<Pagination current={7} total={113} onChange={onChange} />);
    const btnStartPage = getByTitle('5 페이지로 이동');

    fireEvent.click(btnStartPage);
    expect(onChange).toBeCalledWith(5);
  });

  it('7 to 11 page', () => {
    const onChange = jest.fn();
    const { getByTitle } = render(<Pagination current={7} total={124} onChange={onChange} />);
    const btnNextPage = getByTitle('11 페이지로 이동');

    fireEvent.click(btnNextPage);
    expect(onChange).toBeCalledWith(11);
  });

  it('zero total', () => {
    const { getByText, queryByText } = render(<Pagination current={1} total={0} />);

    getByText('1');
    expect(queryByText('2')).toBeNull();
  });
});
