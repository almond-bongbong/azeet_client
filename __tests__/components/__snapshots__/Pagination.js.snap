// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination matches snapshot 1`] = `
<div>
  <div
    class="Pagination__PaginationStyle-sc-1a60n7z-0 kKVOrJ"
  >
    <button
      class="arrow"
      title="첫 페이지로 이동"
      type="button"
    >
      <i
        class="fas fa-angle-double-left"
      />
    </button>
    <button
      class="arrow"
      title="1 페이지로 이동"
      type="button"
    >
      <i
        class="fas fa-angle-left"
      />
    </button>
    <span
      class="numbers"
    >
      <button
        class="num Pagination__NumberStyle-sc-1a60n7z-1 cxuSWD"
        type="button"
      >
        <span>
          1
        </span>
      </button>
      <button
        class="num Pagination__NumberStyle-sc-1a60n7z-1 VFjQJ"
        type="button"
      >
        <span>
          2
        </span>
      </button>
      <button
        class="num Pagination__NumberStyle-sc-1a60n7z-1 VFjQJ"
        type="button"
      >
        <span>
          3
        </span>
      </button>
    </span>
    <button
      class="arrow"
      title="3 페이지로 이동"
      type="button"
    >
      <i
        class="fas fa-angle-right"
      />
    </button>
    <button
      class="arrow"
      title="마지막 페이지로 이동"
      type="button"
    >
      <i
        class="fas fa-angle-double-right"
      />
    </button>
  </div>
</div>
`;
